oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/InMemoryOAuth2AuthorizedClientService.java

```diff
@@ -29,26 +29,40 @@
  * {@link OAuth2AuthorizedClient Authorized Client(s)} in-memory.
  *
  * @author Joe Grandja
+ * @author Vedran Pavic
  * @since 5.0
  * @see OAuth2AuthorizedClientService
  * @see OAuth2AuthorizedClient
  * @see ClientRegistration
  * @see Authentication
  */
 public final class InMemoryOAuth2AuthorizedClientService implements OAuth2AuthorizedClientService {
-	private final Map<String, OAuth2AuthorizedClient> authorizedClients = new ConcurrentHashMap<>();
+	private final Map<String, OAuth2AuthorizedClient> authorizedClients;
 	private final ClientRegistrationRepository clientRegistrationRepository;
 
 	/**
 	 * Constructs an {@code InMemoryOAuth2AuthorizedClientService} using the provided parameters.
 	 *
+	 * @param authorizedClients the map of authorized clients
 	 * @param clientRegistrationRepository the repository of client registrations
 	 */
-	public InMemoryOAuth2AuthorizedClientService(ClientRegistrationRepository clientRegistrationRepository) {
+	public InMemoryOAuth2AuthorizedClientService(Map<String, OAuth2AuthorizedClient> authorizedClients,
```

Comment:
If this constructor is called with a non-empty `Map<String, OAuth2AuthorizedClient> authorizedClients` and the `key(s)` are not correctly built as per `InMemoryOAuth2AuthorizedClientService.getIdentifier()` than a subsequent call to `loadAuthorizedClient()` would return `null` which would be in error. How can we ensure the `key(s)` in the provided `Map` are built correctly?

Reply:
Two options come to mind - either validate keys in constructor, or define a dedicated type for key.

Reply:
Validate keys in constructor would be easy. Can you expand on

> dedicated type for key

Do you mean `Map<SomeNewType, OAuth2AuthorizedClient>` OR an interface that generates the key, for example `KeyGenerator`?

Reply:
I meant the former - having something like `OAuth2AuthorizedClientId` feels like the optimal solution from the perspective of API quality IMO. It encapsulates what in reality we have here - a composite id. At first sight, ensuring backwards compatibility should be doable with `default` methods. I can take a closer look at that as a part of this PR, if you agree.

Reply:
@rwinch Can you expand on this?

> I'd be in favor of creating a static key class that accepts the ClientRegistration as a constructor argument and provides the implementation of the hashCode and equals methods for the key

Sounds like you are in favour of a key class as @vpavic has suggested, for example `OAuth2AuthorizedClientId`? ...that implements `equals` and `hashCode`

Reply:
Yes.

Reply:
> provides the implementation of the hashCode and equals methods for the key vs exposing implementation details of how the key is constructed.

This makes sense to me as well - hiding the implementation details on how the key is constructed using `equals` and `hashCode`. @vpavic Would you like to update this PR based on the discussed approach?

Reply:
Thanks, both - I'll try to update the PR some time during this week.